[🚀 Feature] [py]: Implement BiDi add/remove_request handler

jkim2492 opened the pull request:
### **User description**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->

Implementation of `add_request_handler` and `remove_request_handler` as described in #13993. These methods directly require BiDi objects

* `network.AddIntercept`
* `network.BeforeRequestSent`
* `network.ContinueRequest`
* `network.RemoveRequest`
* `browsingContext.Navigate`

and other added objects are dependencies of objects listed above.


### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
#13993

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->


___

### **PR Type**
Enhancement, Tests


___

### **Description**
- Implemented BiDi network interception and request handling, including adding and removing request handlers.
- Introduced base classes for BiDi objects with JSON serialization and deserialization capabilities.
- Added browsing context navigation classes and network management for remote WebDriver.
- Integrated network management into WebDriver, providing a `network` property for accessing network features.
- Added tests to verify the functionality of BiDi network request handlers, including multiple intercepts and their removal.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>bidi.py</strong><dd><code>Define base classes for BiDi objects</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/selenium/webdriver/common/bidi/bidi.py

<li>Introduced base classes for BiDi objects.<br> <li> Implemented JSON serialization and deserialization methods.<br> <li> Added <code>BidiEvent</code> and <code>BidiCommand</code> classes.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-9bd86ff0b85fa472e4a162953d72219c7eea1802ef8aa2e656078a1d52979f0f">+59/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>browsing_context.py</strong><dd><code>Add browsing context navigation classes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/selenium/webdriver/common/bidi/browsing_context.py

<li>Added <code>Navigate</code> and <code>NavigateParameters</code> classes.<br> <li> Defined <code>BrowsingContext</code> and <code>ReadinessState</code> types.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-a65d02c951aeb477672aa52e5eb0106645ba869aa186c0af6d2672c42cac95c6">+42/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>network.py</strong><dd><code>Implement network interception and request handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/selenium/webdriver/common/bidi/network.py

<li>Implemented classes for network interception and request continuation.<br> <li> Added support for URL patterns and intercept phases.<br> <li> Introduced <code>Network</code> class for managing network operations.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-d04fc5573acd6ca5bcc8167940f134f1a3ead8a5230c066c077a1319a1ffe1d4">+213/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>script.py</strong><dd><code>Extend script module with stack trace classes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/selenium/webdriver/common/bidi/script.py

<li>Added <code>StackFrame</code> and <code>StackTrace</code> classes.<br> <li> Extended existing script functionalities with BiDi objects.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-c4d141b056a48aff05840758a7591f2d2b86d6c2f6b8f79c7362ddc8b41c7fcc">+14/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>network.py</strong><dd><code>Implement network management for remote WebDriver</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/selenium/webdriver/remote/network.py

<li>Implemented network management for remote WebDriver.<br> <li> Added methods for adding and removing request handlers.<br> <li> Introduced context management for network operations.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-46fd53c2cf5a4d87c5ff47cdf75bfecf2e06d5d7888a7c7b1394187c398684e3">+125/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>webdriver.py</strong><dd><code>Integrate network management into WebDriver</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/selenium/webdriver/remote/webdriver.py

<li>Integrated network management into WebDriver.<br> <li> Added <code>network</code> property for accessing network features.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-c87c150ed2a08f5293db58a06da8bf7f14c3a5582586c6270c1be661dfdd0985">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>bidi_network_tests.py</strong><dd><code>Add tests for BiDi network request handlers</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/test/selenium/webdriver/common/bidi_network_tests.py

<li>Added tests for BiDi network request handlers.<br> <li> Verified multiple intercepts and their removal.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-97c8d35b88402ef5de934c8fb95da5f7c7c04039a79baf20cc8fd6c91379dac7">+57/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/SeleniumHQ/selenium?pullRequest=14738) <br/>All committers have signed the CLA.

shbenzer commented:
@AutomatedTester What do you think?

AutomatedTester commented on py/test/selenium/webdriver/common/bidi_network_tests.py:
```diff
@@ -0,0 +1,57 @@
+
+@pytest.mark.xfail_firefox
+@pytest.mark.xfail_safari
+@pytest.mark.xfail_edge
```
Is this not working against Firefox and Edge?

jman-sketch replied:
It was working with edge. I've pushed a commit to remove `xfail_edge`.

As for firefox, it fails with "unsupported operation" at `network.continueRequest`. I see mentions of `network.continueRequest` at https://wiki.mozilla.org/Firefox/Meeting/9-July-2024 , but I recall experiencing a similar problem with firefox when testing #14345.

jkim2492 replied:
It was working with edge. I've pushed a commit to remove `xfail_edge`.

As for firefox, it fails with "unsupported operation" at `network.continueRequest`. I see mentions of `network.continueRequest` at https://wiki.mozilla.org/Firefox/Meeting/9-July-2024 , but I recall experiencing a similar problem with firefox when testing #14345.

jman-sketch commented:
@AutomatedTester 

> Things that I think we need to think about (and I don't think it should block this) is how can we simplify the migration for folks using Classic to BiDi.

Yup. I agree.

jman-sketch commented:
@AutomatedTester  

I also wanted to point out a potential naming issue for BiDi objects. 

Our implementations of BiDi objects (including this PR) include  `class <module name>` inside the respective modules to hold function definitions 

* [Example for "Script"](https://github.com/SeleniumHQ/selenium/blob/83d52f8ba2f32662d028db7ee0373ec048a44e8f/py/selenium/webdriver/common/bidi/script.py#L25). 
* [Example from this PR](https://github.com/SeleniumHQ/selenium/blob/cf3e45927e336397a7a9854afbf301169cd8b7e4/py/selenium/webdriver/common/bidi/network.py#L200)
 
However, some BiDi modules contain types that have the same name as the module which creates a naming conflict.

* [Spec for browsingContext.Browsingcontext](https://w3c.github.io/webdriver-bidi/#type-browsingContext-Browsingcontext)
* [Implementation for browsingContext.Browsingcontext](https://github.com/SeleniumHQ/selenium/blob/cf3e45927e336397a7a9854afbf301169cd8b7e4/py/selenium/webdriver/common/bidi/browsing_context.py#L24)

In the particular case of `Browsingcontext = str`, we can workaround this issue by substituting references to `Browsingcontext` with `str`, but this is not consistent, and also not applicable for more complex types.

jkim2492 commented:
@AutomatedTester 

> Things that I think we need to think about (and I don't think it should block this) is how can we simplify the migration for folks using Classic to BiDi.

Yup. I agree.

jkim2492 commented:
@AutomatedTester  

I also wanted to point out a potential naming issue for BiDi objects. 

Our implementations of BiDi objects (including this PR) include  `class <module name>` inside the respective modules to hold function definitions 

* [Example for "Script"](https://github.com/SeleniumHQ/selenium/blob/83d52f8ba2f32662d028db7ee0373ec048a44e8f/py/selenium/webdriver/common/bidi/script.py#L25). 
* [Example from this PR](https://github.com/SeleniumHQ/selenium/blob/cf3e45927e336397a7a9854afbf301169cd8b7e4/py/selenium/webdriver/common/bidi/network.py#L200)
 
However, some BiDi modules contain types that have the same name as the module which creates a naming conflict.

* [Spec for browsingContext.Browsingcontext](https://w3c.github.io/webdriver-bidi/#type-browsingContext-Browsingcontext)
* [Implementation for browsingContext.Browsingcontext](https://github.com/SeleniumHQ/selenium/blob/cf3e45927e336397a7a9854afbf301169cd8b7e4/py/selenium/webdriver/common/bidi/browsing_context.py#L24)

In the particular case of `Browsingcontext = str`, we can workaround this issue by substituting references to `Browsingcontext` with `str`, but this is not consistent, and also not applicable for more complex types.

shbenzer commented:
@jkim2492 Thank you for your contributions so far! It looks like you've still got a few tests failing - do you know the cause? 

I'm actually the author of the other Python BiDi implementation PR we have up (still failing tests, the browser is hanging during intercept removal iirc): #14592. I'd love to get one of our prs working soon, so I'd be happy to discuss with you any questions/concerns you might have. I haven't had the time to dive back into it recently, so if you know where #14592 is going wrong, or would like help with yours, let me, @p0deje, or @AutomatedTester know.

Also, the team has decided that Request and Response objects are the preferred practice for this functionality, so that'll be a change you'll have to make if you can get yours working - you can take a look at #14592 for more details.

jkim2492 commented:
@shbenzer Thank you!

After adjusting for 875f0a61dfa3b2dafc50a87d91f81204bb51f21a, a lot of the errors cleared up. I'll try to address the other issues soon!

jkim2492 commented:
@p0deje @AutomatedTester @shbenzer I understand that the team has decided to use Response/Request objects for the functions. However, I just wanted to open a discussion about using [RequestData](https://w3c.github.io/webdriver-bidi/#type-network-RequestData) from the BiDi specifications instead and extending it with Selenium's custom Request object "outside".

I believe sticking to the specifications will give us a systematic way to implement BiDi features and also a way to programmatically generate the BiDi objects. What are your thoughts?

shbenzer commented:
Could you provide a sample code block so we can visualize your suggestion?

jkim2492 commented:
In my opinion, we should put objects directly generated from the BiDi objects in `common/bidi` similarly to how we did for cdp.  [Example of generated object](https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-d04fc5573acd6ca5bcc8167940f134f1a3ead8a5230c066c077a1319a1ffe1d4R160-R169). Then in `remote/network.py`, we can wrap around the objects for our convenience.

For instance, the `handler`(callback) parameter of [add_request_handler](https://github.com/SeleniumHQ/selenium/pull/14738/files#diff-46fd53c2cf5a4d87c5ff47cdf75bfecf2e06d5d7888a7c7b1394187c398684e3R88-R90) function in this PR expects a parameter of `RequestData` type, but we can implement our own `Request` type in `remote/network.py` to be used instead

shbenzer commented:
I don’t see a problem with your suggestion. 

The end goal is to have easily understood abstractions for the user - and I think that the Request, Response, and Network objects accomplish that. Your suggestion makes sense, and follows a standard we’re already using somewhere else without sacrificing that end goal. It also follows the BiDi docs more closely, which never hurts.

@p0deje thoughts?

p0deje commented:
We can proceed with any implementation, but we need to make sure it follows a few requirements:

1. It should work w/o async/await. The current version in `script.py` is something I hacked with threads to avoid forcing users to migrate to async implementations. We should keep the same approach.
2. The public API in request/response handling should be wrapped in `Request` / `Response`, which provides means to inspect the objects (e.g. check headers), mutate the requests, and methods to continue/fail. 

Whatever implementation is done internally (data classes for each BiDi spec event, plain dictionaries with data passed around) is not so important. I'm only concerned about the _public interface_.

@shbenzer Your PR tests cause the browser to hang because the thread where the intercepted request should be handled crashes:

```
$ bazel test --test_output=streamed --pin_browsers //py:common-chrome-bidi-test/selenium/webdriver/common/bidi_network_tests.py
...
ERROR    websocket:_logging.py:77 error from callback <function WebSocketConnection._start_ws.<locals>.on_message at 0x105173d30>: 'str' object has no attribute 'from_json'
```

@jkim2492 I would be happy to merge this PR as the foundational pieces are there, but this needs:
1. Get rid of async/await.
2. Implement public API per #13993 (can be done in a follow-up PRs though).

I would also be happy to discuss this in Slack if that would be easier.

jkim2492 commented:
@p0deje I've switched to `WebSocketConnection` and removed  async/awaits. Also confirmed that local tests are passing with
```
bazel test //py:common-chrome-bidi --headless -t-
```
I do prefer this approach to trio.

I would be happy to implement the public API per #13993 in follow up PRs.

Please let me know if there are further steps needed before the merge!

jkim2492 commented:
Weird, a lot of the tests failed due to `apt update` failing. Can't imagine it being caused by changes of this PR

shbenzer commented:
`//dotnet/test/common:BiDi/Storage/StorageTest-edge                       FAILED in 2 out of 2 in 23.3s`
This is unlikely to be related to your pr

p0deje commented on py/selenium/webdriver/remote/webdriver.py:
```diff
@@ -1249,6 +1251,18 @@ def script(self):
+
+        if not self._network:
+            # Bidi 'network.setCacheBehavior' is not implemented in v130
+            self.execute_cdp_cmd("Network.setCacheDisabled", {"cacheDisabled": True})
```
We should never be calling CDP commands from BiDi as the former are only available in Chromium.

jkim2492 replied:
You are right. Initially I had disabled cache to prevent cached responses from bypassing intercepts, but now I think if the user wishes to prevent that behaviour, they should explicitly disable caching

p0deje commented on py/selenium/webdriver/remote/websocket_connection.py:
```diff
@@ -131,7 +132,7 @@ def _process_message(self, message):
             params = message["params"]
             for callback in self.callbacks.get(message["method"], []):
-                callback(params)
+                Thread(target=callback, args=(params,)).start()
```
@shbenzer looks like this is why you had tests hanging!

p0deje commented on py/selenium/webdriver/common/bidi/network.py:
```diff
@@ -0,0 +1,213 @@
+class AddInterceptParameters(BidiObject):
+    phases: typing.List[InterceptPhase]
+    contexts: typing.Optional[typing.List[browsing_context.BrowsingContext]] = None
+    urlPatterns: typing.Optional[typing.List[UrlPattern]] = None
```
Is there any way to use snake_case for BidiObject fields? I guess it would be more idiomatic in Python.

jkim2492 replied:
Possible indeed, although I would prefer to handle it in a followup PR

p0deje commented on py/selenium/webdriver/remote/network.py:
```diff
@@ -0,0 +1,94 @@
+from selenium.webdriver.common.bidi.session import session_unsubscribe
+
+
+class Network:
```
This is our public API. It should only expose the following methods per #13993:
- add_request_handler
- remove_request_handler
- clear_request_handlers

The rest of the methods should be private.

jkim2492 replied:
Making other methods should be straightforward. Is it still okay to add other classes to [py/selenium/webdriver/remote/network.py](https://github.com/SeleniumHQ/selenium/pull/14738/files/47904f7af5b31380d8a3581e78e8365f31b84311#diff-46fd53c2cf5a4d87c5ff47cdf75bfecf2e06d5d7888a7c7b1394187c398684e3)? As mentioned in https://github.com/SeleniumHQ/selenium/pull/14738#issuecomment-2746273983, I would prefer to keep only bidi objects in `common/bidi`, similarly to what we do for CDP

p0deje replied:
My point is that we should not expose any other methods as public API in this module. It's okay to move them to `common/bidi` though.

p0deje commented on py/test/selenium/webdriver/common/bidi_network_tests.py:
```diff
@@ -0,0 +1,55 @@
+    pattern1 = [UrlPatternString(url1)]
+    pattern2 = [UrlPatternString(url2)]
+
+    def request_handler(params):
```
It is probably not clearly stated here, but the idea is to expose only `request` in handler. When it's called, the user can perform the following operations on the request:

1. Mutate the properties of the request (e.g. url, headers, etc)
2. Call `request.continue()` which will proceed with browser processing the request (optionally, with mutated properties).
3. OR call `request.fail()` which tells browser the request errors.

An example of the request object is https://github.com/SeleniumHQ/selenium/blob/1dd967ea4904ff46a8e2ef502084ea8af601c17d/rb/lib/selenium/webdriver/bidi/network/intercepted_request.rb.

cnemo-cenic replied:
Hey, @p0deje, sorry to jump in like this, but I am hoping to use this API (or the other PR) and I wanted to confirm that the body of the request can be read in the request handler. That is what the Ruby class implies, although it is harder to determine for the proposed Python APIs. Is it correct that I can intercept the body then cancel/fail the request?

jkim2492 replied:
@p0deje Thank you for the feedbacks!. In the description above, what should happen if there are multiple `(intercept, callback)` tuples matching a single URL? Would it be one of

- Only the callback that is registered first is called
- All callbacks are called sequentially; where calling `request.continue` moves on to the next callback: until the last callback calls continue which will finally tell the browser to continue.

p0deje replied:
@jkim2492 The preliminary conclusion is that for now, we are saying it's non-deterministic and the order of callbacks is not guaranteed. It's up to the user to ensure there are no duplicated interception callbacks for the same  request.

jkim2492 commented:
Looks like tests that previously failed could have had been due to misconfigured test setup. Tests were passing locally with
```
bazel test //py:common-firefox-bidi-test/selenium/webdriver/common/bidi_network_tests.py -t-
```

jkim2492 commented:
Since #14592 was merged, closing this PR
